[eugenetangkj] iP

eugenetangkj opened the pull request:
## BooBot
> _"Time is the scarcest resource and unless it is managed nothing else can be managed."_  - Peter Drucker ([Source](https://www.azquotes.com/author/4147-Peter_Drucker/tag/time-management#:~:text=Time%20is%20the%20scarcest%20resource,nothing%20else%20can%20be%20managed.&text=Until%20we%20can%20manage%20time%2C%20we%20can%20manage%20nothing%20else.&text=One%20cannot%20buy%2C%20rent%20or%20hire%20more%20time.))

BooBot helps you to keep track of your tasks, saving time and ensuring that you are on track. Problems would be scared away with good task management provided by BooBot.

Main features of BooBot:
- Text-based UI
- Three types of tasks: _Todo_, _Deadline_ and _Event_
- Add and delete tasks from task list.

To get started with BooBot:
1. Install Java.
2. Download the source code [here](https://github.com/nus-cs2103-AY2223S2/ip).
3.  Run the BooBot application.

It takes only 3 steps! Isn't it easy? And not to mention, BooBot is entirely **free**! 😆 


Also, if you are a Java programmer, you can even use the source code to customise BooBot according to your needs! Here is the 
`main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
What are you waiting for? Download BooBot now!
- [x] Download BooBot
- [x] Save your tasks
- [ ] Say bye to your worries!

Commit:
Add Gradle support

eugenetangkj committed:
Changed name of bot from Duke to boo.

eugenetangkj committed:
Completed Level 1: Bot greets user, echoes user's commands, and exits when user types bye.

eugenetangkj committed:
Completed Level 2: Bot stores the text that user enters, and displays them when requested.

eugenetangkj committed:
Added a new Task class.

eugenetangkj committed:
Completed Level 3: Included features to mark tasks as done/undone, and added a new Task class.

eugenetangkj committed:
Made Task an abstract class and created ToDo child class.

eugenetangkj committed:
Created Deadline child class.

eugenetangkj committed:
Created Event class.

eugenetangkj committed:
Implemented feature to add tasks that are todo, deadline or event.

eugenetangkj committed:
Added content into input.txt and expected.txt, and tested using given script.

eugenetangkj committed:
Created DukeException class and handled the mark command

eugenetangkj committed:
Handled exceptions for unmark command.

eugenetangkj committed:
Handled exceptions for the todo command.

eugenetangkj committed:
Handled exceptions for deadline command.

eugenetangkj committed:
Added exception handling for all available commands.

eugenetangkj committed:
Added delete feature to delete tasks.

eugenetangkj committed:
Tidy up the code according to Java code convention

Code does not follow standard Java conventions.

Let's,
* add empty line between description and params in Javadoc
* insert double indentation for lines that overflow
* break up lines that exceed 120 characters.

Following a standard code standard makes it easy for fellow coders
to read and understand the code.

eugenetangkj committed:
Add feature to create file to save task history

Feature that saves task history will make it usable for users,
because users do not need to retype all their tasks if they restart
the bot.

Let's,
* check if data file exists before bot accepts commands
* read from the data file if it exists
* create data file if it does not exists

Checking is required because exceptions will occur if we assume that
a data file exists, which is not true for new users.

eugenetangkj committed:
Add auto-saving feature for task history

Creation of file to store task history needs to be coupled with saving
of task history when the user changes the task list.

Let's add a feature that writes the contents of the user's task list
to the data file whenever the user makes changes to the list.

In the event that the program abruptedly terminated, prior changes to
the task list before termination would still have been saved.

eugenetangkj committed:
Encapsulate date/time in either LocalDateTime or LocalDate object.

Previously, we processed date and time as raw strings so the user could
have entered invalid values.

Let's
* Process user input for date and time and encapsulate them in
the Task classes as either LocalDateTime or LocalDate objects.
* Format the date and time to be more reader-friendly to users.

Ensures that the user types in a valid date and time.

eugenetangkj committed:
Add a help command

With a greater number of commands, it is difficult to remember all
the available commands. It is worse for new users of the chatbot.

Add a help command that will display the list of available commands.

User can now easily view all available commands if he is unsure of what
to type into the chatbot.

eugenetangkj committed:
Add test to ensure start is before end for Event

Need to validate that start is before end, otherwise duration
for an Event will not be valid.

Let's check the LocalDateTime and/or LocalDate objects to check
if user entered a valid duration.

The LocalDateTime and/or LocalDate objects provide API to compare
date and time.

eugenetangkj committed:
Add command to check tasks on a specific day.

Provide users a way to focus on important tasks and narrow down
the search.

Let's add a on date command. Will display all tasks that occur on the
given date. Loop through the ArrayList and process the dates.

Have to check what the type of Task is, and carry out date processing
according to the correct type.

eugenetangkj committed:
Add straight lines between error messages

To be consistent with the style of the bot.

Add two print line statements between exception handling.

In the event of exception, will be handled and message is placed
between 2 straight lines.

eugenetangkj committed:
Modify input.txt and expected.txt

New feature of date and time needs to be tested.

Let's change the input and expected output to conduct regression
testing.

Tests old features and new date/time feature.

eugenetangkj committed:
Merge branch 'branch-Level-7'
Include feature to save task history

eugenetangkj committed:
Merge branch 'branch-Level-8'

Resolve conflicts with Level 8

Could not reconstruct Task objects from Temporal objects.

Let's store the raw strings so we can reconstruct them when reading
from the data file.

Temporal objects can be created using the raw strings, and use them
to create Task objects.

eugenetangkj committed:
Create classes for each command

Each command behaves differently, having its own behaviour.

Let's abstract out the code from Duke and place the behaviour in
each command class.

Easier to identify the code for command behaviour and customise
accordingly.

eugenetangkj committed:
Refactor code into respective classes

Previously, most functionality was in the Duke class, but they do not
represent methods that should belong to Duke bot.

Let's refactor and create classes. Place respective methods in the
correct class.
* Create Ui class to handle input and output.
* Create Storage class to handle saving and loading from local file.
* Create Parser class to handler input parsing.
* Create Command classes to handle command execution.
* Create DateTime class to handle date time processing.
* Create TaskList class to handle operations on the task list.

Placing respective methods in the correct classes allow us to easily
extend and understand our existing program, adhering to OOP principles.

eugenetangkj committed:
Remove unnecessary import statements from Task.java

Imported classes were not being used.

Let's remove the import statements.

Avoid naming conflicts.

eugenetangkj committed:
Merge remote-tracking branch 'upstream/add-gradle-support'
Add Gradle wrapper files for Gradle support.

eugenetangkj committed:
Set up folders for JUnit test with Gradle

Unit testing is not automated.

Let's update build.gradle file and create test folders according to
package structure.

Allows for automation of unit testing by writing JUnit tests in the
respective classes.

eugenetangkj committed:
Create JUnit tests

Unit testing required to test each method independently.

Let's
* Create JUnit tests for DateTime, Parser, Task and ToDo.
* Create TaskListStub class to test Parser independently from TaskList.

JUnit integrates well with Gradle and allows for automated unit
testing. Stub classes required to remove dependencies between classes.

eugenetangkj committed:
Set version of first JAR as v0.1

Keep track of version numbers of released versions of chatbot.

Let's set archiveVersion as v0.1 in build.gradle.

Easily track versions of chatbot.

eugenetangkj committed:
Tidy up JavaDoc comments for classes and methods

Some comments were not properly marked with double aesterisks.

Let's
* Ensure JavaDoc comments are properly marked.
* Add code tag to indicate Java class names in Javadoc comments.

Proper documentation for classes and methods.

eugenetangkj committed:
Align case statements with switch

Some were not aligned.

Let's remove indentations for case within switch blocks.

Adhere to Java coding standards.

eugenetangkj committed:
Add command to find tasks using key phrase.

A long task list makes it difficult for user to narrow down the search
for specific tasks.

Let's add a command to display tasks whose name contains a given key
phrase by iterating through the tasks and looking at each of their name.

Allows us to print out the tasks whose name contains the key phrase.

eugenetangkj committed:
Merge branch 'branch-A-JavaDoc'

eugenetangkj committed:
Merge branch 'branch-A-JavaDoc'
Check that code adheres to coding standard.

eugenetangkj committed:
Add new line between import and class name

Consistency with standards.

Let's add a new line between import and class name.

Follow standards.

eugenetangkj committed:
Adhere to coding standards.

eugenetangkj committed:
Merge branch 'branch-Level-9'
Add feature to find tasks given a key phrase.

eugenetangkj committed:
Attempt to run checkstyle

eugenetangkj committed:
Fix coding standard violations from CheckStyle

There are multiple coding standard violations across the files when
checked using static analyser CheckStyle.

Fixing them would improve readability and consistency.

Let's fix the violations raised by CheckStyle.

eugenetangkj committed:
Merge branch 'A-CheckStyle'
Merge changes that prevent coding standard violations.

eugenetangkj committed:
Update gradle file for JavaFX

eugenetangkj committed:
Add echo gui and start command refactor

eugenetangkj committed:
Create GUI for chat bot

Previously, there was no GUI and user had to interact with the bot
via CLI.

Create a GUI would improve the aesthetics of the application.

Let's use JavaFX to create a GUI, and for all Command classes, return
the String output to be processed instead of using System.out.println.

eugenetangkj committed:
Restructure code for GUI, using FXML

Previously, code for GUI is contained within the Duke class, with
the attributes being customised within a single method. This is
not scalable when application becomes larger as the method would
become too long.

Let's factor out the code into new classes and keep information of the
controls in FXML files instead. Then, use SceneBuilder to allow further
customisation such as CSS rules.

This makes it easier to expand the scale of the application when more
UI elements are added and abstract away the UI details from the Duke
class. Also, customisation is simpler.

eugenetangkj committed:
Merge branch 'branch-Level-10'
Change from text-based CLI to GUI input

VietAnh1010 commented on src/main/java/duke/command/ByeCommand.java:
```diff
@@ -0,0 +1,28 @@
+
+
+    /**
+     * Constructs a <code>ByeCommand</code>.
```
You can instead use `{@code ByeCommand}` - it will be more concise.

eugenetangkj replied:
Okay, thanks for the suggestion!

VietAnh1010 commented on src/main/java/duke/datetime/DateTime.java:
```diff
@@ -0,0 +1,127 @@
+     * @param end The Temporal encapsulating the end date and time.
+     * @return true if both refer to the same day, else false.
+     */
+    public static boolean isEqualDate(Temporal start, Temporal end) {
```
Good naming for a method that returns a boolean.

VietAnh1010 commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,206 @@
+    private String data_file_folder_path;
+
+    /** Absolute file path where the data file should be stored in. **/
+    private String data_file_path;
```
This naming style does not follow the convention. Can you explain the reason behind this?

eugenetangkj replied:
Sorry, my bad. Must have been doing too much Python...

VietAnh1010 commented on src/main/java/duke/tasklist/TaskListStub.java:
```diff
@@ -0,0 +1,20 @@
+
+
+    /**
+     * Return the value 3.
```
Should be `Returns` instead of `Return`.

eugenetangkj replied:
You are right!

VietAnh1010 reviewed (commented):
Overall, your code looks good. There are little to no style errors.

eugenetangkj committed:
Reduce size of GUI window

Previous GUI window is too large for laptop screen resolutions, making
the text field and button to be out of screen and cannot be clicked on.

Let's reduce the size of the GUI and arrange the components to make it
fit in smaller laptop screens.

eugenetangkj committed:
Add assertions for sanity check

There are parts where execution should not reach and where it should
only reach once.

Let's add assertion statements to check the control flow.

Provides a sanity check to ensure to detect possible control flow
errors.

eugenetangkj committed:
Try update build.gradle

eugenetangkj committed:
Modify build.gradle to enable Assertions

Previously, assertions were not enabled.

Let's set enableAsssertions to be true.

Allows for assertions to run when executing commands.

liujiayue314 commented on src/main/java/duke/command/EventCommand.java:
```diff
@@ -0,0 +1,71 @@
+        StringBuilder sb = new StringBuilder();
+        sb.append("Added task to list:\n");
+        sb.append(newEventTask.getStatusOfTaskInString() + "\n");
+        if (tasks.getSizeOfTaskList() == 1) {
```
nice to see that u even took care of singular and plural nouns

liujiayue314 commented on src/main/java/duke/command/OnCommand.java:
```diff
@@ -0,0 +1,73 @@
+/**
+ * Represents an on command that is entered by the user to check what tasks are there on a specific day.
+ */
+public class OnCommand extends Command {
```
Maybe can think of a better name, maybe a verb, that is more description of the action.

eugenetangkj replied:
Alright! Thanks for the suggestion. 👍

liujiayue314 reviewed (commented):
Overall very good work with detailed javadoc and easy to read code.

eugenetangkj committed:
Updated code tags

eugenetangkj committed:
Change names of files from Duke to Boo

eugenetangkj committed:
Remove unused methods

eugenetangkj committed:
Refactor the parseRawCommand method in Parser

Previously, all parsing and checking of validity was done in the
parseRawCommand method. This means that the method is very long and
contains too much logic.

Let's extract out the logic for validating each type of command into
its own method, and call the corresponding validation method depending
on which command it is.

This simplifies the parseRawCommand method, and makes code neater as
it is easier to find the logic that validates each type of command,
and the method will not grow much larger whenever we add a new command
to the bot.

eugenetangkj committed:
Simplify code of DataTime by removing complex conditionals

eugenetangkj committed:
Refine code quality

eugenetangkj committed:
Merge pull request #1 from eugenetangkj/branch-A-Assertions

Add assertions

eugenetangkj committed:
Rename DukeException to BooException

eugenetangkj committed:
Resolve conflict in Command

eugenetangkj committed:
Remove duke folder

eugenetangkj committed:
Squash bug regarding assertion

An assertion was previously inserted such that it caused the program
to crash. Predicted logic was incorrect

Let's remove the assertion.

The assertion was placed in a block where execution should flow
multiple times instead of just once. Hence, it is redundant.

eugenetangkj committed:
Merge pull request #2 from eugenetangkj/branch-A-CodeQuality

Rename of bot and improve code quality

eugenetangkj committed:
Set up continous integration

Previously, continous integration was not set up so commits or updates
to PR will not run any workflows.

Let's add a workflow via a .yml file that will carry out actions
whenever an update is made.

Allows checking to be done when one commits to the central repository.

eugenetangkj committed:
Select windows platform

eugenetangkj committed:
Fix checkstyle violations

eugenetangkj committed:
Fix style error for ParserTest method name

eugenetangkj committed:
Create ReminderCommand handling in Parser

Parser does not currently handle reminder commands.

Let's create a method to validate reminder commands, and parse it
accordingly.

Allows us to cater to the reminder commands when user types in raw
input.

eugenetangkj committed:
Add reminder feature

Currently, application does not support providing reminders to users.

Let's add a reminder feature by creating a ReminderCommand which
displays deadlines and events within a specified duration from the
current date and time.

This allows users to easily view upcoming deadlines and events, so
they can prioritise tasks accordingly.

eugenetangkj committed:
Update build.gradle for change of app name

eugenetangkj committed:
Fix checkstyle errors in classes

eugenetangkj committed:
Merge branch 'branch-B-Reminders'
Add new reminder feature

eugenetangkj committed:
Fix format for Javadocs of CommandType

eugenetangkj committed:
Refactor runCommand for reminderCommand

eugenetangkj committed:
Refactor validateDeadline of Parser

eugenetangkj committed:
Refactor validateEvent of Parser

eugenetangkj committed:
Change window icon

eugenetangkj committed:
Add screenshot of application

eugenetangkj committed:
Start on Readme

eugenetangkj committed:
Modify user guide

Previous user guide did not explain features of the application.

Let's include explanation of each feature and provide additional
information such as quick start and command summary.

This makes the user guide intuitive and user-friendly so that a new
user can easily learn how to use the application.

eugenetangkj committed:
Update README

eugenetangkj committed:
Fix bug regarding find and on command

eugenetangkj committed:
Fix typo in user guide for event command

eugenetangkj committed:
Add config.yml file

There is inconsistency on how a table is displayed for GitHub file
preview and GitHub pages.

Let's add a config.yml file to change the Markdown processor to ensure
consistency between GitHub file preview and GitHub pages.

Allows a table to be rendered correctly on GitHub pages.

eugenetangkj committed:
Fix typos and add acknowledgement links in user guide

eugenetangkj committed:
Fix extra word in user guide

eugenetangkj committed:
Add minor changes to readme

eugenetangkj committed:
Insert link to user guide in README

eugenetangkj committed:
Modify link to user guide in README

eugenetangkj committed:
Update README.md
